HDDS-6646. Intermittent failure in TestOzoneFileSystem#testRenameToTrashEnabled

devmadhuu opened the pull request:
## What changes were proposed in this pull request?

This test case was flaky with listStatus issue which got resolved as part of [HDDS-9041](https://issues.apache.org/jira/browse/HDDS-9041)

Following points taken care as part of this PR:

1. testRenameToTrashEnabled : which the current patch fixes.
2. testTrash : this patch improves it, but still marked flaky as there is still a possibility of failure in some cases. will be handled as part of [HDDS-6645](https://issues.apache.org/jira/browse/HDDS-6645)
3. deleteRootDir cleanup method has been updated to intermittent failures of running test class with some other test cases. This change is not directly impacting `testRenameToTrashEnabled`

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-6646

## How was this patch tested?

This patch was tested with 10 times job runs with existing test cases.

hemantk-12 commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFileSystem.java:
```diff
@@ -802,6 +802,7 @@ protected void deleteRootDir() throws IOException, InterruptedException {
       return;
     }
     deleteRootRecursively(fileStatuses);
+    Thread.sleep(500);
```
Can we use [GenericTestUtils#waitFor](https://github.com/apache/ozone/blob/387c53781aae03f944906a0add58f51250d36687/hadoop-hdds/test-utils/src/main/java/org/apache/ozone/test/GenericTestUtils.java#L214) instead of `Thread.sleep()`?

devmadhuu replied:
> Can we use [GenericTestUtils#waitFor](https://github.com/apache/ozone/blob/387c53781aae03f944906a0add58f51250d36687/hadoop-hdds/test-utils/src/main/java/org/apache/ozone/test/GenericTestUtils.java#L214) instead of `Thread.sleep()`?

Thank you @hemantk-12 for review. I have handled the comment.

devmadhuu replied:
> Can we use [GenericTestUtils#waitFor](https://github.com/apache/ozone/blob/387c53781aae03f944906a0add58f51250d36687/hadoop-hdds/test-utils/src/main/java/org/apache/ozone/test/GenericTestUtils.java#L214) instead of `Thread.sleep()`?

@hemantk-12 thanks for the review, I tried using GenericTestUtils, however this in actual run throwing always TimeOutException and not working in this case.

hemantk-12 replied:
I'm surprise that `Thread.sleep()` works fine but [GenericTestUtils#waitFor](https://github.com/apache/ozone/blob/387c53781aae03f944906a0add58f51250d36687/hadoop-hdds/test-utils/src/main/java/org/apache/ozone/test/GenericTestUtils.java#L214) doesn't. Just curious what was lambda used?

I feel this should work if it is just about cleaning dir.
```
    GenericTestUtils.waitFor(() -> {
      try {
        FileStatus[] fileStatus = fs.listStatus(ROOT);
        return fileStatus != null && fileStatus.length == 0;
      } catch (IOException e) {
        return false;
      }
    }, 100, 500);
    ```

devmadhuu replied:
Thanks @hemantk-12 for review. I have handled as per your suggestion.

hemantk-12 reviewed (commented):
Thanks @devmadhuu for fixing it.

Can you please also add how this patch would fix the issue in the PR description?

devmadhuu commented:
> Thanks @devmadhuu for fixing it.
> 
> Can you please also add how this patch would fix the issue in the PR description?

This patch directly alone not contributing the resolution of this issue, however another [patch](https://github.com/apache/ozone/pull/5204) which has fixed issue in listStatus API fixed the root cause as this test also was showing intermittent fail in deleteRootDir only, but when running multiple times in multiple jobs, this looks to be a race condition when tests running in parallel. And after adding sleep, I can see list status showing correct results.

devmadhuu commented:
@sumitagrawl Pls review.

sadanand48 commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFileSystem.java:
```diff
@@ -1618,15 +1621,13 @@ public void testGetTrashRoots() throws IOException {
    */
   @Test
-  @Flaky("HDDS-6646")
   public void testRenameToTrashEnabled() throws Exception {
```
This particular test never calls deleteRootDir(), not sure how fixing deletRootDir makes  this test non-flaky, Anyway I feel this functionality is tested in testTrash method in the same class in much detail . This test was written before the trash feature was complete, I feel we can remove this test from the class.

devmadhuu replied:
Thanks Sadanand for review. As discussed, we need to test Trash related tests in separate Test class spinning separate mini cluster having  emptier thread started with close match of timings of relevant  assertion statements and corresponding trash and checkpoint interval. So I have moved assertion for above test in testTrash test case and will be handled as part of [HDDS-6645](https://issues.apache.org/jira/browse/HDDS-6645)

sadanand48 replied:
Thanks Devesh, sounds good to me.

sumitagrawl commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFileSystem.java:
```diff
@@ -795,13 +796,15 @@ public void testListStatusOnKeyNameContainDelimiter() throws Exception {
       return;
     }
     deleteRootRecursively(fileStatuses);
+    Thread.sleep(500);
```
How sleep for 500ms solves the problem?
And Who throws TimeoutException in what case make it flaky ?

hemantk-12 replied:
+1

I don't get either why `TimeoutException` is added [here](https://github.com/apache/ozone/pull/5217/files#diff-0a2a989278da5b28bd0346a9db79c98ca319ff238eb150af17e2686afabfc1e5R800). May be you forgot to remove it while reverting it.

devmadhuu replied:
Since now GenericTestUtils is being used, I have not removed TimeoutException. This change is not directly related, however many other tests are getting impacted because of deleteRootDIr cleanup, so this is important as part of this test as well.

sumitagrawl reviewed (commented):
@devmadhuu thanks for working over this, have few query over the changes.

hemantk-12 commented:
> > Thanks @devmadhuu for fixing it.
> > Can you please also add how this patch would fix the issue in the PR description?
> 
> This patch directly alone not contributing the resolution of this issue, however another [patch](https://github.com/apache/ozone/pull/5204) which has fixed issue in listStatus API fixed the root cause as this test also was showing intermittent fail in deleteRootDir only, but when running multiple times in multiple jobs, this looks to be a race condition when tests running in parallel. And after adding sleep, I can see list status showing correct results.

Can you please add the same in the description?

hemantk-12 commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFileSystem.java:
```diff
@@ -795,15 +796,19 @@ public void testListStatusOnKeyNameContainDelimiter() throws Exception {
-    Thread.sleep(500);
     fileStatuses = fs.listStatus(ROOT);
+    FileStatus[] finalFileStatuses = fileStatuses;
+    GenericTestUtils.waitFor(
```
You are supposed to fetch `fs.listStatus(ROOT);` inside the lambda.

https://github.com/apache/ozone/pull/5217#discussion_r1308026768

devmadhuu commented:
@sumitagrawl @hemantk-12 @sadanand48  Kindly re-review.

sadanand48 commented:
Thanks @devmadhuu , patch looks good to me,  IIUC, there are 3 tests in question:
1. `testRenameToTrashEnabled` :  which the current patch fixes.
2. `testTrash` : this patch improves it, but still marked flaky as there is still a possibility of failure in some cases. will be handled as part of HDDS-6645
3. `testListStatusOnLargeDirectory` : which by fixing deleteRoot by adding a wait should stabilise the test further.

hemantk-12 commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFileSystem.java:
```diff
@@ -1627,19 +1633,19 @@ public void testRenameToTrashEnabled() throws Exception {
+    // We can safely assert only trash directory here.
+    // Asserting Current or checkpoint directory is not feasible here in this
+    // test due to independent TrashEmptier thread running in cluster and
+    // possible flakyness is hard to avoid unless we test this test case
```
```suggestion
    // possible flakiness is hard to avoid unless we test this test case
```

hemantk-12 commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFileSystem.java:
```diff
@@ -1657,15 +1663,21 @@ public void testTrash() throws Exception {
 
+    // Call moveToTrash. We can't call protected fs.rename() directly
+    trash.moveToTrash(path);
+    // Added this assertion here and will be tested as part of testTrash
```
I'm not following this comment. What do you mean by `will be tested as part of testTrash`? This is `testTrash` test. Is there another `testTrash`? What am I missing?

hemantk-12 reviewed (commented):
Thanks for addressing review comments.

Overall changes are fine. 

Few doubts and suggestions:
1. PR title is incorrect I think. It is supposed to be `HDDS-6646 Intermittent failure ....`
2. Can you please share the workflow run where you run the test 10 times? I found [this](https://github.com/devmadhuu/ozone/actions/runs/5938962749/job/16109360981), but in last iteration test still failed. Also I would suggest to run only  `testRenameToTrashEnabled` and have a green CI for it.
3. I think [deleteRootDir](https://github.com/apache/ozone/blob/011de37b19ec874819ae5bbf726580ba5b0b7f93/hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFileSystem.java#L721) is not needed here because it is deleted in [@After](https://github.com/apache/ozone/blob/011de37b19ec874819ae5bbf726580ba5b0b7f93/hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFileSystem.java#L228). If you agree, remove this either as part of this PR or may be separate PR.

devmadhuu commented:
> Thanks for addressing review comments.
> 
> Overall changes are fine.
> 
> Few doubts and suggestions:
> 
> 1. PR title is incorrect I think. It is supposed to be `HDDS-6646 Intermittent failure ....`
> 2. Can you please share the workflow run where you run the test 10 times? I found [this](https://github.com/devmadhuu/ozone/actions/runs/5938962749/job/16109360981), but in last iteration test still failed. Also I would suggest to run only  `testRenameToTrashEnabled` and have a green CI for it.
> 3. I think [deleteRootDir](https://github.com/apache/ozone/blob/011de37b19ec874819ae5bbf726580ba5b0b7f93/hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFileSystem.java#L721) is not needed here because it is deleted in [@After](https://github.com/apache/ozone/blob/011de37b19ec874819ae5bbf726580ba5b0b7f93/hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFileSystem.java#L228). If you agree, remove this either as part of this PR or may be separate PR.

@hemantk-12 , we can change PR title, however it's better to keep with JIRA title.

I'll re-run the test in multiple runs once again and get green CI.

I think we should keep deleteRootDir change as i observed that without this change, CI will not pass. Many other PR are blocked because of this

hemantk-12 commented:
> > Thanks for addressing review comments.
> > Overall changes are fine.
> > Few doubts and suggestions:
> > 
> > 1. PR title is incorrect I think. It is supposed to be `HDDS-6646 Intermittent failure ....`
> > 2. Can you please share the workflow run where you run the test 10 times? I found [this](https://github.com/devmadhuu/ozone/actions/runs/5938962749/job/16109360981), but in last iteration test still failed. Also I would suggest to run only  `testRenameToTrashEnabled` and have a green CI for it.
> > 3. I think [deleteRootDir](https://github.com/apache/ozone/blob/011de37b19ec874819ae5bbf726580ba5b0b7f93/hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFileSystem.java#L721) is not needed here because it is deleted in [@After](https://github.com/apache/ozone/blob/011de37b19ec874819ae5bbf726580ba5b0b7f93/hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFileSystem.java#L228). If you agree, remove this either as part of this PR or may be separate PR.
> 
> @hemantk-12 , we can change PR title, however it's better to keep with JIRA title.
> 
> I'll re-run the test in multiple runs once again and get green CI.
> 
> I think we should keep deleteRootDir change as i observed that without this change, CI will not pass. Many other PR are blocked because of this

1. I didn't mean to remove Jira ID or Jira Title. I meant `HDDS-6646. Intermittent failure in TestOzoneFileSystem#testRenameToTrashEnabled`  not `Hdds 6646 - Intermittent failure in TestOzoneFileSystem#testRenameToTrashEnabled`. PR title is supposed to start with `HDDS-JiraID.` followed by PR or jIra title
2. I am curious how `deleteRootDir()` calling solves the issue in test [testListStatusOnLargeDirectory](https://github.com/apache/ozone/blob/011de37b19ec874819ae5bbf726580ba5b0b7f93/hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFileSystem.java#L720). Did you try to remove it and run? Do you have a run where it failed after it was removed?

devmadhuu commented:
> > > Thanks for addressing review comments.
> > > Overall changes are fine.
> > > Few doubts and suggestions:
> > > 
> > > 1. PR title is incorrect I think. It is supposed to be `HDDS-6646 Intermittent failure ....`
> > > 2. Can you please share the workflow run where you run the test 10 times? I found [this](https://github.com/devmadhuu/ozone/actions/runs/5938962749/job/16109360981), but in last iteration test still failed. Also I would suggest to run only  `testRenameToTrashEnabled` and have a green CI for it.
> > > 3. I think [deleteRootDir](https://github.com/apache/ozone/blob/011de37b19ec874819ae5bbf726580ba5b0b7f93/hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFileSystem.java#L721) is not needed here because it is deleted in [@After](https://github.com/apache/ozone/blob/011de37b19ec874819ae5bbf726580ba5b0b7f93/hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFileSystem.java#L228). If you agree, remove this either as part of this PR or may be separate PR.
> > 
> > 
> > @hemantk-12 , we can change PR title, however it's better to keep with JIRA title.
> > I'll re-run the test in multiple runs once again and get green CI.
> > I think we should keep deleteRootDir change as i observed that without this change, CI will not pass. Many other PR are blocked because of this
> 
> 1. I didn't mean to remove Jira ID or Jira Title. `...` just mean the continuation. I meant `HDDS-6646. Intermittent failure in TestOzoneFileSystem#testRenameToTrashEnabled`  not `Hdds 6646 - Intermittent failure in TestOzoneFileSystem#testRenameToTrashEnabled`. PR title is supposed to start with `HDDS-JiraID.` followed by PR or jIra title
> 2. I am curious how `deleteRootDir()` calling solves the issue in test [testListStatusOnLargeDirectory](https://github.com/apache/ozone/blob/011de37b19ec874819ae5bbf726580ba5b0b7f93/hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFileSystem.java#L720). Did you try to remove it and run? Do you have a run where it failed after it was removed?

1. I fixed the PR title. Thanks for pointing out.
2. `testListStatusOnLargeDirectory` test was failed like any other test due to deleteRootDIr cleanup. No changes being done except deleteRootDir, May be we can remove from PR description about this test.
3.

devmadhuu commented:
Here is the successful green CI from fork [run](https://github.com/devmadhuu/ozone/actions/runs/6024556982) in fork.

hemantk-12 reviewed (approved):
Thanks for the patch @devmadhuu. 

Overall looks good to me.

> [deleteRootDir](https://github.com/apache/ozone/blob/011de37b19ec874819ae5bbf726580ba5b0b7f93/hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFileSystem.java#L721) is not needed in [testListStatusOnLargeDirectory](https://github.com/apache/ozone/blob/011de37b19ec874819ae5bbf726580ba5b0b7f93/hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFileSystem.java#L720) because it is deleted in [@After](https://github.com/apache/ozone/blob/011de37b19ec874819ae5bbf726580ba5b0b7f93/hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFileSystem.java#L228).

I still believe there is no need to call `deleteRootDir` in `testListStatusOnLargeDirectory`. Because it is not expected from HDDS-6646, it can be handle as a separate task.

sadanand48 commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFileSystem.java:
```diff
@@ -1623,35 +1615,6 @@ public void testGetTrashRoots() throws IOException {
-   */
-  @Test
-  @Flaky("HDDS-6646")
-  public void testRenameToTrashEnabled() throws Exception {
```
Was the test removal intended? because earlier commit had fixed the test and description also says so.

devmadhuu replied:
> Was the test removal intended? because earlier commit had fixed the test and description also says so.

Yes @sadanand48 , as [HDDS-6645](https://issues.apache.org/jira/browse/HDDS-6645) will be taking care of testTrash test case fix, so the only assertion which was different from `testRenameToTrashEnabled` and `testTrash` has been moved to  `testTrash`  and `testRenameToTrashEnabled` has been removed as per suggestion also from @sumitagrawl

smengcl commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFileSystem.java:
```diff
@@ -802,20 +802,12 @@ protected void deleteRootDir() throws IOException, InterruptedException {
-    // Waiting for double buffer flush before calling listStatus() again
-    // seem to have mitigated the flakiness in cleanup(), but at the cost of
-    // almost doubling the test run time. M1 154s->283s (all 4 sets of params)
-    cluster.getOzoneManager().awaitDoubleBufferFlush();
```
I'm not sure the removal of this line belongs to this PR?

#5244 still looks required.

devmadhuu replied:
Yes @smengcl #5244 needs anyway. I think while resolving conflict, it got removed.

smengcl replied:
Ok. 5 out of 9 commits merged to master branch after this PR starts to have flaky `integration (filesystem)` again as a result of this removal. Need to merge #5244 asap.

https://github.com/apache/ozone/actions/runs/6100351514/job/16554482439
https://github.com/apache/ozone/actions/runs/6105400581/job/16568864195
https://github.com/apache/ozone/actions/runs/6107352731/job/16574193577
https://github.com/apache/ozone/actions/runs/6110706683/job/16584346159
https://github.com/apache/ozone/actions/runs/6114248488/job/16595437359

smengcl commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFileSystem.java:
```diff
@@ -1623,35 +1615,6 @@ public void testGetTrashRoots() throws IOException {
-   */
-  @Test
-  @Flaky("HDDS-6646")
-  public void testRenameToTrashEnabled() throws Exception {
```
Would you update the PR description to reflect that `testRenameToTrashEnabled` is removed for now?

hemantk-12 replied:
+1

devmadhuu replied:
@smengcl @hemantk-12 , Thanks for noticing that. Updated the PR description.
